netty/src/main/java/io/grpc/netty/QueuedCommand.java

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.netty;
+
+import com.google.common.base.Objects;
+
+import io.grpc.Status;
+import io.grpc.internal.AbstractStream2;
+import io.grpc.internal.ClientTransport.PingCallback;
+import io.grpc.internal.Stream;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.util.Recycler;
+
+import java.util.concurrent.Executor;
+
+/**
+ * Simple wrapper type around a command and its optional completion listener.
+ */
+interface QueuedCommand {
+
+  /**
+   * Returns the promise beeing notified of the success/failure of the write.
+   */
+  ChannelPromise promise();
+
+  /**
+   * Sets the promise.
+   */
+  void promise(ChannelPromise promise);
+
+  /**
+   * Returns this command to the command pool.
+   */
+  void recycle();
+
+  static final class UnionCommand implements QueuedCommand, CancelClientStreamCmd,
```

Comment:
Single recycler is correct.  It also means not having to call the specific recycler from each context.